### DISCUSSION TITLE
Managing Hyas integrations via Node.js packages (npm) vs. Hugo modules: Pros and Cons

### DISCUSSION THREAD

**Discussion by salim-b** [gethyas/hyas#420] [Ideas]
I want to learn what all the (perceived) pros and cons of using Node.js packages (npm) instead of "native" [Hugo modules](https://gohugo.io/hugo-modules/) are. I understand that Hyas uses npm to make Hugo itself a dependency of the website you're building (using [hugo-installer](https://github.com/dominique-mueller/hugo-installer/)). But Hyas integrations like [images](https://github.com/gethyas/images), [inline-svg](https://github.com/gethyas/inline-svg) and [seo](https://github.com/gethyas/seo) could in principle still be added as Hugo modules instead of npm packages.

(Perceived) pros of npm I imagine:

- Hugo modules require a working installation of Golang. This can be worked around to some part by [`hugo mod vendor`](https://gohugo.io/commands/hugo_mod_vendor/)ing the modules into the project with the added downside of significantly increasing the project's Git repo size (although it's debatable how much a downside this actually is when only using a dozen or less modules that have little to no transitive dependencies). 

  npm package dependencies in a Hugo module are also not a problem. [Hugo 0.75](https://gohugo.io/news/0.75.0-relnotes/) introduced [`hugo mod npm pack`](https://gohugo.io/commands/hugo_mod_npm_pack/) to create a composite `package.json` that accounts for all npm dependencies used in imported Hugo modules. So it's still possible to offer a convenient out-of-the-box experience via `npm install` for Hyas theme end-users.

(Perceveived) cons of npm I can think of:

- Not embracing Hugos native dependency management infrastructure makes maintaining Hyas integrations more work, I guess. Also users have to wait for new npm releases.

  With Hugo modules, we just reference Git refs – as soon as the referenced branch is updated, users can fetch the updated code via [`hugo mod get`](https://gohugo.io/hugo-modules/use-modules/#update-modules) if they like. [KISS](https://en.wikipedia.org/wiki/KISS_principle).

Did I miss misinterpret something? Other thoughts?

**Comment by h-enk** [comment]
Some of my findings why I didn't opt for using Hugo Modules:

- A lot of Hugo Modules also include npm modules, thereby introducing a *second* system for managing dependencies
- There's no global registry of available packages for Hugo Modules, like there is for [npm](https://www.npmjs.com/)
- You cant' manage *versions* with Hugo Modules, you'll always get the latest
- You can't manage *scripts* in Hugo Modules, like you can with npm
- `hugo mod npm pack [flags] [args]` is experimental and it failed my tests for Hyas. Also note that *scripts* are not supported
- Hugo Modules require Go, introducing another dependency

Some of my findings on why I did opt for using npm as a dependency manager:

- One system for managing dependencies
- Dependencies are explicitly declared in a single place
- Installing and updating are handled by npm
- Your project is locked onto specific versions

**Comment by h-enk** [comment]
> I understand that Hyas uses npm to make Hugo itself a dependency of the website you're building (using [hugo-installer](https://github.com/dominique-mueller/hugo-installer/)).

Hugo is usually installed in a global location. This way however, your project users could be using *different* Hugo versions. Hyas however, installs Hugo *locally* in your project. This way, your project users will always use the *same* Hugo version. Also now, you and your project users will always use a Hugo version that *works* with (is tested for) a specific version of Hyas.

**Reply by salim-b** [reply]
Thank you very much for sharing your thoughts! 

I have no real intention of lobbying for Hugo modules, I'm just interested in the technical debate to better understand and potentially improve the whole Hyas setup (which is rather complex, but quite awesome!).

In the following I'll reply to some of your points:

> There's no global registry of available packages for Hugo Modules, like there is for [npm](https://www.npmjs.com/)

Indeed. An official Hugo modules catalogue would improve discoverability tremendously. But on a purely technical level, we don't need a registry for them, public Git forges like GitHub suffice. And there is no actual advantage in the Hyas integrations being on npm regarding discoverability, it appears to me...

> You cant' manage versions with Hugo Modules, you'll always get the latest

This is not quite true, [you can reference specific Git tags ("versions")](https://gohugo.io/hugo-modules/use-modules/#get-a-specific-version) as well as specific *commits*, see e.g. [this forum thread](https://discourse.gohugo.io/t/how-to-point-to-a-specific-commit-in-my-fork-of-a-hugo-module/35720) for more information.

> `hugo mod npm pack [flags] [args]` is experimental and it failed my tests for Hyas.

This is valuable information! I might run some tests myself in the future, to confirm this.

> Hugo is usually installed in a global location. This way however, your project users could be using _different_ Hugo versions. Hyas however, installs Hugo _locally_ in your project. This way, your project users will always use the _same_ Hugo version. Also now, you and your project users will always use a Hugo version that _works_ with (is tested for) a specific version of Hyas.

There is the [`hugoVersion` setting](https://gohugo.io/hugo-modules/configuration/#module-configuration-hugoversion) in Hugo's module config which essentially serves the same purpose *without* duplicating the Hugo binary for every project. Of course, people have to install the appropriate Hugo version themselves, which could be project-local if they like. There's also a nice CLI called [**`hvm`**](https://github.com/jmooring/hvm) that essentially allows for an evenly convenient UX as hugo-installer (just check in the `.hvm` file into a Git repo) but with using a single copy of the Hugo binary *per Hugo version* instead of *per Hugo project*, thus reducing disk space requirements considerably if you have a lot of Hugo sites.

---

tl;dr: I agree to

> Hugo Modules require Go, introducing another dependency

and think this is the most important downside of using Hugo modules vs. npm, since we need npm anyways. 🙂

**Reply by h-enk** [reply]
> and think this is the most important downside of using Hugo modules vs. npm, since we need npm anyways. 🙂

Yes, I agree, and in addition to that, one of the core design principles for Hyas is to make managing dependencies and scripts as easy as possible.

I think a *single* system for managing dependencies — like [npm](https://www.npmjs.com/) for Node.js (or  [Composer](http://getcomposer.org/) for PHP, [Bundler](http://bundler.io/) for Ruby, or [pip](https://pip.pypa.io/en/stable/) for Python for that matter)  — is more in line with that principle than a dual system like Hugo Modules.

**Reply by salim-b** [reply]
Well, I don't think it is that important to do everything with a *single* dependency system – as long as it doesn't introduce additional work for end-users. The Hyas base being an npm package but managing its *integrations* as Hugo modules would be completely transparent for end-users[^note], while at the same time being more convenient for developers. You could also share your work on Hyas integrations with the broader Hugo community. For example, there is already [a Hugo module for Bootstrap (5)](https://github.com/gohugoio/hugo-mod-bootstrap-scss), no need for [a Hyas-specific integration](https://github.com/gethyas/bootstrap) anymore.

That being said, I'm totally fine with Hyas continuing to use npm exclusively as long as you're happy to maintain all those npm packages. 😄

[^note]: As mentioned before, the modules could even be *vendored*, so Hyas would also work out-of-the-box for users that don't have Golang installed.

**Reply by h-enk** [reply]
> The Hyas base being an npm package but managing its integrations as Hugo modules would be completely transparent for end-users[1](https://github.com/orgs/gethyas/discussions/420#user-content-fn-note-98bac864ea6b3955bac8d7b32fca5c54), while and at the same time being more convenient for developers.

I see at least two types of users.

Hyas is a web framework using Hugo as the build engine and npm for managing dependencies (including Hugo itself). You could look at Hyas as a bridge between the Javascript world and the Hugo world. From a Javascript world perspective, npm is the most logical choice for managing dependencies. From a Hugo world perspective that is Hugo Modules with (experimental) support for npm.

To me — as a hybrid — the Hugo Modules way is fairly complex (and not robust) and npm is way more user-friendly (and more widely accepted). I also like that npm [explicitly declares and isolates dependencies](https://12factor.net/dependencies), whereas Hugo Modules does that more implicitly. 

For now, I'm still in favor of an npm-only solution (I guess because I like npm more and like to keep things simple). However, it's an interesting thought if you could use Hugo Modules next to that — if you'd want that. We'll have to set up a POC to see how that works.

**Reply by h-enk** [reply]
> However, it's an interesting thought if you could use Hugo Modules next to that — if you'd want that. We'll have to set up a POC to see how that works.

Well, that was easy — what's not to love about Hugo?!

- Repository: https://github.com/h-enk/hyas-hugo-mod-bootstrap
- Website: https://main--steady-pony-d10294.netlify.app/

What I did:
- I started with [template-basic-starter-rec](https://github.com/gethyas/create-hyas/tree/main/template-basic-starter-rec)
- Then added [hugo-mod-bootstrap-scss](https://github.com/gohugoio/hugo-mod-bootstrap-scss)
- Updated [themes/my-basic-theme/assets/scss/app.scss](https://github.com/h-enk/hyas-hugo-mod-bootstrap/blob/main/themes/my-basic-theme/assets/scss/app.scss)
- And, booom!

Good to know (and document) that this is a possibility — if you fancy that.

**Reply by salim-b** [reply]
Nice, glad to hear your PoC works as expected!

**Comment by salim-b** [comment]
Another important piece of information:

Multiple `mounts` within the same Hugo module (incl. the top-level project) with `target = "assets"` [are not guaranteed to work reliably](https://discourse.gohugo.io/t/how-to-inspect-hugos-virtual-union-filesystem/46190/8), i.e. the overlaying seems random[^doc]. Future improvements to Hugo's mounts behaviour might improve on this situation according to Hugo's lead dev, but currently it's pretty-much broken. ~~I just realized that the exact same mounts config produces different results with `npm run dev`/`hugo server` than with `npm run build`/`hugo`/Netlify: With the former, the `assets/scss/layouts/_custom.scss` file from the top-level project wins, but with with the latter it's the one from the doks-core node module.~~

In the course of trying to understand how Hugo mounts work I seem to have pissed off Hugo's lead dev – he [closed the forum topic without further comments](https://discourse.gohugo.io/t/how-to-inspect-hugos-virtual-union-filesystem/46190/10), so I couldn't ask more specifically. But apparently Hugo does merge `assets` correctly between `imports` but not necessarily between multiple `mounts` in the same module.

***Update**: The cause of the different results of `hugo server` vs. `hugo` was not Hugo modules but environments (`development` vs. `production`) and an outdated PurgeCSS config in the latter environment which didn't whitelist the `common/_custom.scss` Bootstrap file. 😑 Sorry for the confusion. Nevertheless I think Hyas should switch to properly set up Hugo modules and **stop** managing Hugo dependencies via npm. Hugo's `assets` component folder merge behaviour is currently a pain, regardless of my false alarm above (striked out).*

[^doc]: It's likely no coincidence that the merge behaviour of `assets` – in contrast to all other relevant component folders – is [*not documented*](https://gohugo.io/hugo-modules/theme-components/) (additionally, that doc hasn't been updated to mention modules; too many Hugo devs really hate properly documenting the awesome shit they throw at us uninitiated fools 🙄).
